File: python/src/m3e/static.py
"""Núcleo linear do M3E estático.

O ordenamento das incógnitas reproduz a planilha Excel:
``[y, pi_expected, pi, r_expected]``.
"""

from __future__ import annotations

from dataclasses import asdict, dataclass, replace
from typing import Mapping

import numpy as np
import pandas as pd


STATIC_VARIABLES = ("y", "pi_expected", "pi", "r_expected")


@dataclass(frozen=True)
class StaticParameters:
    """Parâmetros e variáveis exógenas do equilíbrio estático."""

    y_potential: float = 1.0
    y_autonomous: float = 1.1
    r_cb: float = 0.05
    pi_long_run: float = 0.035
    pi_target: float = 0.035
    gamma: float = 2.0
    theta: float = 0.3
    alpha: float = 0.25
    phi_pi: float = 1.5

    def __post_init__(self) -> None:
        if self.gamma <= 0:
            raise ValueError("gamma deve ser estritamente positivo")
        if not 0 <= self.theta < 1:
            raise ValueError("theta deve pertencer ao intervalo [0, 1)")
        if self.alpha <= 0:
            raise ValueError("alpha deve ser estritamente positivo")
        if self.phi_pi < 0:
            raise ValueError("phi_pi não pode ser negativo")

    @property
    def r_natural(self) -> float:
        return (self.y_autonomous - self.y_potential) / self.gamma

    def with_changes(self, **changes: float) -> "StaticParameters":
        return replace(self, **changes)

    def as_dict(self) -> dict[str, float]:
        values = asdict(self)
        values["r_natural"] = self.r_natural
        return values


@dataclass(frozen=True)
class StaticSolution:
    """Solução, matriz inversa e diagnósticos de um equilíbrio."""

    parameters: StaticParameters
    coefficients: np.ndarray
    exogenous: np.ndarray
    inverse: np.ndarray
    values: np.ndarray

    def as_dict(self) -> dict[str, float]:
        return dict(zip(STATIC_VARIABLES, self.values, strict=True))

    def as_series(self) -> pd.Series:
        return pd.Series(self.as_dict(), name="equilibrium")

    @property
    def output_gap(self) -> float:
        return float(self.values[0] - self.parameters.y_potential)

    @property
    def nominal_rate_ex_ante(self) -> float:
        return float(self.values[3] + self.values[1])

    @property
    def nominal_rate_ex_post(self) -> float:
        return float(self.values[3] + self.values[2])

    @property
    def residual(self) -> np.ndarray:
        return self.coefficients @ self.values - self.exogenous


def _coerce_parameters(
    parameters: StaticParameters | Mapping[str, float] | None,
) -> StaticParameters:
    if parameters is None:
        return StaticParameters()
    if isinstance(parameters, StaticParameters):
        return parameters
    return StaticParameters(**dict(parameters))


def static_matrices(
    parameters: StaticParameters | Mapping[str, float] | None = None,
) -> tuple[np.ndarray, np.ndarray]:
    """Constrói a matriz de coeficientes ``A`` e o vetor exógeno ``b``.

    O sistema é ``A @ x = b``, com
    ``x = [y, pi_expected, pi, r_expected]``.
    """

    p = _coerce_parameters(parameters)
    coefficients = np.array(
        [
            [1.0, 0.0, 0.0, p.gamma],
            [0.0, 1.0, -p.theta, 0.0],
            [-p.alpha, -1.0, 1.0, 0.0],
            [0.0, -p.phi_pi, 0.0, 1.0],
        ],
        dtype=float,
    )
    exogenous = np.array(
        [
            p.y_autonomous,
            (1.0 - p.theta) * p.pi_long_run,
            -p.alpha * p.y_potential,
            p.r_cb - p.phi_pi * p.pi_target,
        ],
        dtype=float,
    )
    return coefficients, exogenous


def solve_static(
    parameters: StaticParameters | Mapping[str, float] | None = None,
) -> StaticSolution:
    """Resolve o equilíbrio estático e devolve todos os objetos matriciais."""

    p = _coerce_parameters(parameters)
    coefficients, exogenous = static_matrices(p)
    inverse = np.linalg.inv(coefficients)
    values = inverse @ exogenous
    return StaticSolution(p, coefficients, exogenous, inverse, values)


def compare_static(
    base: StaticSolution | StaticParameters | Mapping[str, float] | None = None,
    scenario: StaticSolution | StaticParameters | Mapping[str, float] | None = None,
) -> pd.DataFrame:
    """Compara dois equilíbrios e calcula desvios absolutos e relativos."""

    base_solution = base if isinstance(base, StaticSolution) else solve_static(base)
    scenario_solution = (
        scenario if isinstance(scenario, StaticSolution) else solve_static(scenario)
    )
    base_values = base_solution.values
    scenario_values = scenario_solution.values
    with np.errstate(divide="ignore", invalid="ignore"):
        relative = np.where(base_values != 0, scenario_values / base_values - 1.0, np.nan)
    return pd.DataFrame(
        {
            "variable": STATIC_VARIABLES,
            "equilibrium_1": base_values,
            "equilibrium_2": scenario_values,
            "absolute_deviation": scenario_values - base_values,
            "relative_deviation": relative,
        }
    )


def excel_static_scenarios() -> tuple[StaticParameters, StaticParameters]:
    """Retorna as duas colunas de parâmetros da planilha estática."""

    equilibrium_1 = StaticParameters()
    equilibrium_2 = equilibrium_1.with_changes(y_autonomous=1.2)
    return equilibrium_1, equilibrium_2

File: R/m3eReplication/R/static.R
M3E_STATIC_VARIABLES <- c("y", "pi_expected", "pi", "r_expected")

.m3e_validate_static <- function(parameters) {
  if (parameters$gamma <= 0) stop("gamma deve ser estritamente positivo")
  if (parameters$theta < 0 || parameters$theta >= 1) {
    stop("theta deve pertencer ao intervalo [0, 1)")
  }
  if (parameters$alpha <= 0) stop("alpha deve ser estritamente positivo")
  if (parameters$phi_pi < 0) stop("phi_pi não pode ser negativo")
  parameters
}

m3e_static_defaults <- function(...) {
  parameters <- list(
    y_potential = 1.0,
    y_autonomous = 1.1,
    r_cb = 0.05,
    pi_long_run = 0.035,
    pi_target = 0.035,
    gamma = 2.0,
    theta = 0.3,
    alpha = 0.25,
    phi_pi = 1.5
  )
  changes <- list(...)
  if (length(changes)) parameters <- modifyList(parameters, changes)
  parameters$r_natural <-
    (parameters$y_autonomous - parameters$y_potential) / parameters$gamma
  .m3e_validate_static(parameters)
}

.m3e_static_parameters <- function(parameters = NULL) {
  defaults <- m3e_static_defaults()
  if (is.null(parameters)) return(defaults)
  parameters$r_natural <- NULL
  result <- modifyList(defaults, parameters)
  result$r_natural <- (result$y_autonomous - result$y_potential) / result$gamma
  .m3e_validate_static(result)
}

m3e_static_matrices <- function(parameters = NULL) {
  p <- .m3e_static_parameters(parameters)
  A <- matrix(
    c(
      1, 0, 0, p$gamma,
      0, 1, -p$theta, 0,
      -p$alpha, -1, 1, 0,
      0, -p$phi_pi, 0, 1
    ),
    nrow = 4,
    byrow = TRUE,
    dimnames = list(
      c("IS", "expectations", "Phillips", "monetary_rule"),
      M3E_STATIC_VARIABLES
    )
  )
  b <- c(
    y_autonomous = p$y_autonomous,
    expectations = (1 - p$theta) * p$pi_long_run,
    Phillips = -p$alpha * p$y_potential,
    monetary_rule = p$r_cb - p$phi_pi * p$pi_target
  )
  list(A = A, b = b, variable_order = M3E_STATIC_VARIABLES)
}

m3e_solve_static <- function(parameters = NULL) {
  p <- .m3e_static_parameters(parameters)
  matrices <- m3e_static_matrices(p)
  A_inverse <- solve(matrices$A)
  values <- drop(A_inverse %*% matrices$b)
  names(values) <- M3E_STATIC_VARIABLES
  result <- list(
    parameters = p,
    A = matrices$A,
    b = matrices$b,
    A_inverse = A_inverse,
    values = values,
    output_gap = values[["y"]] - p$y_potential,
    nominal_rate_ex_ante = values[["r_expected"]] + values[["pi_expected"]],
    nominal_rate_ex_post = values[["r_expected"]] + values[["pi"]],
    residual = drop(matrices$A %*% values - matrices$b)
  )
  class(result) <- "m3e_static_solution"
  result
}

.m3e_as_static_solution <- function(value) {
  if (inherits(value, "m3e_static_solution")) return(value)
  m3e_solve_static(value)
}

m3e_compare_static <- function(base = NULL, scenario = NULL) {
  if (is.null(base) || is.null(scenario)) {
    defaults <- m3e_excel_static_scenarios()
    if (is.null(base)) base <- defaults$equilibrium_1
    if (is.null(scenario)) scenario <- defaults$equilibrium_2
  }
  base_solution <- .m3e_as_static_solution(base)
  scenario_solution <- .m3e_as_static_solution(scenario)
  relative <- ifelse(
    base_solution$values != 0,
    scenario_solution$values / base_solution$values - 1,
    NA_real_
  )
  data.frame(
    variable = M3E_STATIC_VARIABLES,
    equilibrium_1 = unname(base_solution$values),
    equilibrium_2 = unname(scenario_solution$values),
    absolute_deviation = unname(scenario_solution$values - base_solution$values),
    relative_deviation = unname(relative),
    check.names = FALSE
  )
}

m3e_excel_static_scenarios <- function() {
  list(
    equilibrium_1 = m3e_static_defaults(),
    equilibrium_2 = m3e_static_defaults(y_autonomous = 1.2)
  )
}
